Added support SOCKS5 proxy and NO_PROXY option

remittor opened the pull request:
Related commit: https://github.com/TortoiseGit/TortoiseGit/commit/dcdd4333426c511b99ea7a3849882d193f4e3c53
Related issue: https://gitlab.com/tortoisegit/tortoisegit/-/issues/3590

This is what the dialog box looks like:

![sh_0930](https://user-images.githubusercontent.com/9350115/96259722-afa99580-0fc6-11eb-9ab1-c457bc98de5d.jpg)

remittor committed:
Added support SOCKS5 proxy

Add new field "Type" to Settings -> Network -> Proxy server

Signed-off-by: remittor <remittor@gmail.com>

remittor committed:
Added support NO_PROXY environment for git

Add new field "No Proxy list" to Settings -> Network -> Proxy server

Signed-off-by: remittor <remittor@gmail.com>

csware commented:
Thank you for your contribution!

We'll have a look in the near future. ATM I'd don't have too much spare time.

csware commented:
Do you have any idea why the tests fail?

csware commented on src/Git/Git.cpp:
```diff
@@ -2291,6 +2291,14 @@ BOOL CGit::CheckMsysGitDir(BOOL bFallback)
+	if (!proxy.IsEmpty()) {
+		CString no_proxy = CRegString(L"Software\\TortoiseGit\\Git\\Servers\\global\\http-proxy-ignore", L"");
+		if (!no_proxy.IsEmpty()) {
+			m_Environment.SetEnv(L"NO_PROXY", no_proxy);
```
With this we seem to have a proxy configuration in git config and also in the environment. I'd prefer to have it only in config ATM.

remittor replied:
The user can change the gitconfig file in the text editor.
Because of this, the settings in the gitconfig file are prioritized.

csware replied:
Enrivonment variables have higher precedence, Therefore, "Because of this, the settings in the gitconfig file are prioritized." is wrong.

remittor replied:
I thought so too before. But source code of `libgit2` shows the opposite:
https://github.com/libgit2/libgit2/blob/b106834d70ae94d054c0127b35572010723fee4f/src/remote.c#L832-L846
Settings from the GIT config have a higher priority.

csware replied:
libgit2 is not used for remote

csware replied:
Ah, there is no "no proxy" config in Git.

remittor replied:
> Enrivonment variables have higher precedence
Look at this: https://github.com/curl/curl/blob/f3b7d72c57ed84189ce6e1ca3426e9681236ddf2/lib/url.c#L2437-L2478
The `curl` uses the `http_proxy` env setting only if there is no `-x` option in the startup parameters.

csware replied:
Maybe a better approach is to set the environment variable at a different place.

remittor replied:
> Maybe a better approach is to set the environment variable at a different place.

Yeah. Calling `GetConfigValue` from the `CGit` constructor was not a good idea.
I had to make a special function `GetPreparedEnvironment` to change the environment.

csware commented:
Please include your real name into the Signed-off-by line and Git meta data.

remittor commented:
> Do you have any idea why the tests fail?

It has nothing to do with my PR.
Bug was clearly present before.

csware commented:
> It has nothing to do with my PR.
> Bug was clearly present before.

How do you know? master tests cleanly.

remittor commented:
> > It has nothing to do with my PR.
> > Bug was clearly present before.
> 
> How do you know? master tests cleanly.

I'll try to check it again tomorrow.

remittor commented:
> Do you have any idea why the tests fail?

This commit is the cause of the error:
https://github.com/TortoiseGit/tgit/commit/73f192c991016bf88a9416cdf0e949f8b946f7e2
Lazy initialization was removed incorrectly! 
After these changes, the dependent code began to work incorrectly:
https://github.com/TortoiseGit/TortoiseGit/blob/364ebc6e686828b24964845bdb52445b645b9e27/ext/gitdll/gitdll.c#L888-L890

```
try {
  GetConfigValue(L"http.proxy")
    CheckAndInitDll
      setup_git_directory_gently(NULL)
        die("not a git repository (or any of the parent directories): %s", DEFAULT_GIT_DIR_ENVIRONMENT)
} catch (...)
{
  /* the_repository->gitdir is NULL !!! */
}

git_get_config(key, CStrBufA(value, 4096), 4096)
  have_git_dir()
  get_git_dir()
    BUG("git environment hasn't been setup");  
```

PR for fix this bug: https://github.com/TortoiseGit/tgit/pull/15

remittor committed:
Dynamic changing environment before starting the GIT

This is better than calling function CheckAndInitDll in the constructor of class CGit.
